File: ModeladoProgramacion/Huffman/huffman.hs
--Modelado y programacion 2019-2
--Tarea 2
--Alumnos: Cruz Pérez Ramón, Emilio Caballero Jimenez
import Data.Char
import Data.List

data Arbol a = Vacio
            | Nodo (String,Int) (Arbol a) (Arbol a)
            deriving (Eq,Ord,Show,Read)

--Devuelve una lista con el char y sus repeticiones
nodos :: String -> [(String, Int)]
nodos [] = []
nodos (x:xs) = [repeticiones [x] xs] ++ nodos (eliminaRepeticiones [x] xs)
-- Elimina el char del String para que no contemos otra vez el mismo caracter
eliminaRepeticiones :: String -> String -> String
eliminaRepeticiones x xs = [y | y <- xs, [y] /= x]
-- Nos da el numero de repeticiones de un char
repeticiones :: String -> String -> (String,Int)
repeticiones x xs = (x, length [y | y <- xs, [y] == x] +1)

--Ordena la lista de menor mayor
ordenaLista :: [(String,Int)] -> [(String,Int)]
ordenaLista [] = []
ordenaLista (x:xs) = ordenaLista(menores) ++ [x] ++ ordenaLista(mayores)
          where
              menores = [y | y <-xs, snd y < snd x]
              mayores = [z | z <-xs, snd z >= snd x]
--Nos regresa la lista con las repeticiones de cada letra ordena
listaDuplas :: String -> [(String,Int)]
listaDuplas s = ordenaLista(nodos s)

--Crea todos los las hojas con su fucion
crearN :: [(String, Int)] -> [Arbol a]
crearN [] = []
crearN [x] = [Nodo (fst x, snd x) (Vacio) (Vacio)]
crearN (x:xs) = [(Nodo ((fst x ++ fst (head xs)), (snd x + (snd (head xs)))) (Nodo(fst x,snd x) (Vacio) (Vacio) ) (Nodo (fst (head xs), snd(head xs)) (Vacio) (Vacio)))] ++ crearN (tail xs)

--da la triplera de los arboles
dupla :: Arbol a -> (String,Int)
dupla (Nodo (x,y) t1 t2)  = (x,y)

--Nos da el arbol completo gracias a crearArbol
arbolHuffman :: String -> Arbol a
arbolHuffman s = head(crearArbol(crearN(listaDuplas(s))))

--- funcion recursiva que crea el arbol completo, con una lista de arboles con la union de las hojas
crearArbol :: [Arbol a] -> [Arbol a]
crearArbol [] = []
crearArbol [x] = [x]
crearArbol (x:xs) = crearArbol ([Nodo ((fst (dupla x)++ fst (dupla (head xs))), (snd (dupla x) + (snd (dupla (head xs)))))  (x) (head xs)]++(tail xs))

--funcion que devuelve una lista con 101010 dependiendo la letra
listaCodigo :: Arbol a -> [(String,String)]
listaCodigo a = listaC (a) ("")
--funcion auxiliar de metodo listaCodigo Devuelve una lista con 01010 dependiendo la letra
listaC :: Arbol a -> String -> [(String,String)]
listaC a n = case a of
                Vacio -> []
                (Nodo a Vacio Vacio) -> [((n), (fst (a)))]
                (Nodo a t1 t2) ->(listaC t1 (n++"0")) ++ (listaC t2 (n++"1"))

--codifica la cadena con ayuda del metodo codificar
codificar :: String -> Arbol a -> String
codificar s a = crearCodigo s (listaCodigo a)
---Funcion que recibe una cadena y una lista para codificarla
crearCodigo :: String -> [(String,String)] ->  String
crearCodigo [] (x:xs) = []
crearCodigo (x:xs) [] = []
crearCodigo (x:xs) (y:ys) = go [x] (y:ys)++ crearCodigo (xs) (y:ys)
---funcion auxiliar para crearCodigo nos ayuda a obtener la cadena de 101010
go :: String -> [(String,String)] ->  String
go [] (x:xs) = []
go (x:xs) [] = []
go (x:xs) (y:ys) = if isSubsequenceOf [x] (snd(y))
                   then fst(y)
                   else crearCodigo (x:xs) ys

--decodifica la cadena de 10100 a cadela original o cualquier otra
decodifica :: String  -> Arbol a -> Arbol a -> String
decodifica [] e (Nodo a Vacio Vacio) = fst(a)
decodifica (x:xs) e a = case a of
                Vacio -> []
                (Nodo a Vacio Vacio) -> fst(a)++ decodifica (x:xs) e e
                (Nodo a t1 t2)-> if isSubsequenceOf [x] "0"
                              then decodifica (xs) e t1
                              else decodifica (xs) e t2

File: ModeladoProgramacion/Huffman/Huffman.java
import java.util.ArrayList;

public class Huffman{
  ArrayList<Nodo> lista = new ArrayList<Nodo>();
  private class Nodo{
    String caracter;
    int peso;
    String valor;
    Nodo padre;
    Nodo izquierdo;
    Nodo derecho;
    Boolean hoja= false;
    public Nodo (){
      caracter = "";
      peso = 0;
      valor = "-1";
      izquierdo = null;
      derecho = null;
    }
    public Nodo(String s, int p){
      caracter = s;
      peso = p;
      valor = "-1";
      izquierdo = null;
      derecho = null;
    }
    public void tostring(){
      String s = caracter+","+peso;
    }
    public String caracter(){
      return caracter;
    }
    public int peso(){
      return peso;
    }
    public int compareTo(Nodo n){
      if (peso < n.peso) {
        return -1;
      }else {
        return (peso == n.peso)? 0:1;
      }
    }

  }


  /**
  **con la cadena que recibe crea el arbol
  */
  public Huffman(String s){
    creaLista(s);
  }

  /**
  **Cuenta el numero de repeticiones de una char
  */
  public int cuenta(char c, String s){
    int cont = 0;
    for (int i =0; i < s.length() ; i++ ) {
        if (c == s.charAt(i)) {
          cont++;
        }
    }
    return cont;
  }

  /**
  *elimina las letras de una String
  */
  public String eliminaString(int cont,char c, String s){
    int j=0;

    while(j != s.length()){//pude que no borre uno
      if (c == s.charAt(j)) {
        s = s.substring(0,j)+ s.substring(j+1);
      }else{
      j++;
      }
    }
    return s;
  }

  /**
  *crea una lista ordena con un cadena (creo las hojas en una listas)
  */
  public void creaLista(String s){
    Nodo n ;
    int i = 0;
      while(!s.equals("")){
        int numero = cuenta(s.charAt(i),s);
        n = new Nodo( Character.toString(s.charAt(i)),numero);
        n.hoja = true;
      //  s= s.replace("a", "");
        s = eliminaString(numero,s.charAt(i),s);
        lista.add(n);
      }
      lista.sort((o1, o2) -> o1.compareTo(o2));
  }

/**
**imprime la lista, sirve para hacer puebas
*/
  public void imprime(){
    for (int i=0; i < lista.size(); i++) {
      System.out.println((lista.get(i)).caracter()+", "+(lista.get(i)).peso());
    }
  }

/**
**Crea el arbol apartir de la lista ordena de las hojas
*/
  public Nodo arbol(){
    Nodo p = new Nodo();
    while(lista.size() != 1){
      Nodo n = (lista.size() >= 1)? lista.get(0):null;
      lista.remove(0);
      Nodo n2 = (lista.size() >= 1)? lista.get(0):null;
      lista.remove(0);
      p = new Nodo();
      String s =  n.caracter + n2.caracter;
      p.caracter = s;
      p.peso = n.peso + n2.peso;
      p.izquierdo = n;
      p.derecho = n2;
      n.valor ="0";
      n.padre = p;
      n2.valor = "1";
      n2.padre = p;
      lista.add(p);
    }
    return lista.get(0);
  }

/**
** recibe la cadena normal  y el arbol hecho
**/
  public String codifica(String s,Nodo n){
    String r  = "";
    Nodo m = n;
    while(s.length() != 0 ){
      if (cuenta(s.charAt(0),(m.izquierdo).caracter) > 0) {
          r += m.izquierdo.valor;
         if(m.izquierdo.hoja){
           s = s.substring(1);
           m = n;
         }else {
           m = m.izquierdo;
         }

      }else if (cuenta(s.charAt(0),(m.derecho).caracter) > 0) {
        r += m.derecho.valor;
        if(m.derecho.hoja){
          s = s.substring(1);
          m = n;
        }else {
          m = m.derecho;
        }

      }
    }
    return r;
  }


  /**
  *decodifica un String de 10101010
  *@return s
  */
  public String decodifica(String s, Nodo n){
    String r  = "";
    Nodo m = n;
    while(s.length() != 0 ){
      if (Character.toString(s.charAt(0)).equals(m.izquierdo.valor)) {
          s = s.substring(1);
         if(m.izquierdo.hoja){
           r += m.izquierdo.caracter;
           m = n;
         }else {
           m = m.izquierdo;
         }
      }else if (Character.toString(s.charAt(0)).equals(m.derecho.valor)) {
        s = s.substring(1);
        if(m.derecho.hoja){
          r += m.derecho.caracter;
          m = n;
        }else {
          m = m.derecho;
        }
      }
    }
    return r;

  }

  public static void main(String[] args) {
    String s = " hola como estass";
    Huffman a = new Huffman(s); // se crea arbol
    String x = a.codifica(s,a.arbol()); //X = 1010101010
    String y = a.decodifica(x,a.arbol());// Y = " hola como estass"
    System.out.println(x);
    System.out.println(y);
  }




}
